nuxeo-features/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/action/RecomputeThumbnailsAction.java

```diff
@@ -0,0 +1,153 @@
+                       .build();
+    }
+
+    public static class RecomputeThumbnailsComputation extends AbstractBulkComputation {
+
+        protected String xpath;
+
+        public RecomputeThumbnailsComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        public void startBucket(String bucketKey) {
+            BulkCommand command = getCurrentCommand();
+            xpath = command.getParam(PARAM_XPATH);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+
+            DocumentModelList docs = loadDocuments(session, ids);
+
+            for (DocumentModel doc : docs) {
+                processDoc(session, doc);
+            }
+        }
+
+        protected void processDoc(CoreSession session, DocumentModel doc) {
+            Blob thumbnailBlob = getManagedThumbnail(doc);
+            if (thumbnailBlob == null) {
+                ThumbnailAdapter thumbnailAdapter = doc.getAdapter(ThumbnailAdapter.class);
+                if (thumbnailAdapter == null) {
+                    return;
+                }
+                thumbnailBlob = thumbnailAdapter.computeThumbnail(session);
+            }
+            if (thumbnailBlob != null) {
+                if (!doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                    doc.addFacet(ThumbnailConstants.THUMBNAIL_FACET);
+                }
+                doc.setPropertyValue(ThumbnailConstants.THUMBNAIL_PROPERTY_NAME, (Serializable) thumbnailBlob);
+            } else {
+                if (doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                    doc.setPropertyValue(ThumbnailConstants.THUMBNAIL_PROPERTY_NAME, null);
+                    doc.removeFacet(ThumbnailConstants.THUMBNAIL_FACET);
+                }
+            }
+            if (doc.isDirty()) {
+                doc.putContextData(VersioningService.VERSIONING_OPTION, VersioningOption.NONE);
+                doc.putContextData(VersioningService.DISABLE_AUTO_CHECKOUT, Boolean.TRUE);
+                doc.putContextData(DublinCoreListener.DISABLE_DUBLINCORE_LISTENER, Boolean.TRUE);
+                doc.putContextData(NotificationConstants.DISABLE_NOTIFICATION_SERVICE, Boolean.TRUE);
+                doc.putContextData("disableAuditLogger", Boolean.TRUE);
+                if (doc.isVersion()) {
+                    doc.putContextData(ALLOW_VERSION_WRITE, Boolean.TRUE);
+                }
+                doc.putContextData(THUMBNAIL_UPDATED, true);
+                session.saveDocument(doc);
+            }
+        }
+
+        protected Blob getManagedThumbnail(DocumentModel doc) {
+            BlobHolder bh = doc.getAdapter(BlobHolder.class);
+            if (bh == null) {
+                return null;
+            }
+            Blob blob = bh.getBlob();
+            if (blob == null) {
+                return null;
+            }
```

Comment:
i think it's nice to have from my point of view :) and reduce the lines of codes , to replace by 
`
if(bh == null || bh.getBlob()){
  return null;
}
`

Reply:
You mean
```
if(bh == null || bh.getBlob() == null)
```
?
But the `blob` variable is used below.

Reply:
yes i l mean bh.getBlob() == null. thx
for blob using i agree, but i think have a small bloc to exit when the required data are null is better. that save some lines code :)

Reply:
`BlobHolder bh = doc.getAdapter(BlobHolder.class);
            if (bh == null || bh.getBlob() == null) {
                return null;
            }
            Blob blob = bh.getBlob();`

Reply:
But `bh.getBlob()` (potentially costly as accesses the backend) is executed twice...